general/mem_manager.hpp

```diff
@@ -907,13 +913,25 @@ inline void Memory<T>::MakeAlias(const Memory &base, int offset, int size)
    h_mt = base.h_mt;
    h_ptr = base.h_ptr + offset;
    if (!(base.flags & REGISTERED))
-   { flags = (base.flags | ALIAS) & ~(OWNS_HOST | OWNS_DEVICE); }
-   else
    {
-      const size_t s_bytes = size*sizeof(T);
-      const size_t o_bytes = offset*sizeof(T);
-      MemoryManager::Alias_(base.h_ptr, o_bytes, s_bytes, base.flags, flags);
+      if (MemoryManager::RegisterAliasBases())
+      {
+         // Register 'base':
+         MemoryManager::Register_(base.h_ptr, nullptr, base.capacity*sizeof(T),
+                                  base.h_mt, base.flags & OWNS_HOST,
+                                  base.flags & ALIAS, base.flags);
+      }
+      else
```

Comment:
and if we are in the case where `host_mem_type == device_mem_type == MANAGED`, shouldn't we register it?

Reply:
The `MANAGED` type is (almost) always registered.

I think there are three cases in which we skip registration: (1) the host pointer has type `MemoryType::HOST`, (2) the host pointer is `NULL`, and (3) the host pointer (with any host memory type) was externally allocated and then wrapped with `Memory::Wrap` without passing ownership to the `Memory` object.

In `Memory::MakeAlias`, we want to register the base if there is any chance that in the future the new alias `Memory` may be accessed on device, i.e. with `MemoryClass::DEVICE`. Generally, we cannot predict the future (😁 ) but we can make informed guesses. If we fail to predict correctly that an alias will be accessed on device, then we should still get the error `Cannot register an alias!`. Currently, the "educated guess" is defined by the method `MemoryManager::RegisterAliasBases` and that makes a guess based on the globally-configured host and device memory types -- specifically "predict future device access" if the types are different.

In the case `host_mem_type == device_mem_type == MANAGED` we make a probably wrong prediction and we do not register the base. In most cases this will be fine since the host type of the base is probably `MANAGED` and it is already registered. There are, however, cases where this prediction will be wrong: any of the 3 cases above where we do not register the host pointer of the base but later we try to access the alias on device (the case of NULL pointer may actually still be okay -- we can check on that if necessary).

To fix some of these "bad predictions", we can change the behavior of `RegisterAliasBases` -- e.g. to return true if the globally configured device memory type is a device type. This seems like a better choice. What do you think?

Note that this choice may also fail in some cases: let's say we run on CPU, but there is TPL that we use that only works on GPU; naturally we will pass the data from a `Vector` to that library by asking the `Vector`'s `Memory` explicitly for a device pointer with a call like `vec.GetMemory().Read(MemoryClass::DEVICE, size)` which will fail if `vec.data` is an alias and it's base was not previously registered.

Reply:
What's the benefit to predict and leave room for errors when we could just register all the time? `Read`, `ReadWrite`, `Write` are not performance critical operations.